### DISCUSSION TITLE
How to feed raw bytes into USBMIDI_Interface?

### DISCUSSION THREAD

**Discussion by progmars** [tttapa/Control-Surface#798] [Q&A] [answered]
I'm working on a device that receives MIDI bytes from another device as raw bytes and has to pass them to USB MIDI as actual MIDI messages. The source is not serial, BLE or any other normally MIDI-compatible input port, therefore Control Surface does not have interfaces for that.

I have started with code like this:

```
USBMIDI_Interface usbMidi;
```

As I see, USBMIDI_Interface does not have any kind of send function that would be able to collect raw bytes until they form a complete MIDI message and only then send it to USB.

So I was wondering what would be the most straight-forward way to create a custom interface that accumulates incoming bytes and then I could call write() and read() in loop() and as soon as I get a valid message, I would pass it to usbMidi.send(). Are there any quick examples or should I try to do something on my own using `Parsing_MIDI_Interface`  as a base class? Or could StreamMIDI_Interface have more complete functionality? It sounds like it - I want to stream bytes into it and get MIDI messages out. But I haven't seen useful examples for StreamMIDI_Interface yet. 

Also, I'm confused about begin(), loop() and update() calls in the examples. Some of them call Control_Surface.begin() and Control_Surface.loop(). But some call usbMidi.begin() and usbMidi.update() instead.

What is the difference? When should what be used?

I know one thing - the loop() needs to read incoming USB MIDI messages, no matter if I use them or not, otherwise some DAWs can overflow the outgoing buffer (there has been a discussion on Teensy forums about this). Judging by the source code, `usbMidi.update();` should be enough to deal with this issue. Will Control_Surface.loop() do the same thing? Or maybe Control_Surface.loop() overkill for my needs and usbMidi.update() should be enough...

I'm using Teensy LC, by the way.

**Comment by progmars** [comment]
I think I finally found something; sharing it in case it might be useful to others. It turns out I don't need the entire interface, I can reuse the Control Surface MIDI parser alone. It's a bit confusingly named SerialMIDI_Parser but actually, it is not linked to a serial port at all, it just accumulates bytes in a serial manner and tries to extract messages.

```
SerialMIDI_Parser parser;

// in the loop:

    // feed the payload bytes and try to parse the message out, if possible
    MIDIReadEvent event = MIDIReadEvent::NO_MESSAGE;

    // in my case, payload is known to always contain a single MIDI message only
    for (uint8_t pl = 0; pl < payload.length; pl ++)
    {
        event = parser.parse(payload.message[pl]);
    }
    payload.usedLength = 0; // to prevent consuming the same data in the next iteration until we receive new bytes in the payload

    // only a single message is expected in the payload, so we can process it outside the "for" loop
    if (event == MIDIReadEvent::CHANNEL_MESSAGE)
    {
        usbMidi.send(parser.getChannelMessage());
    } 
    else if (event == MIDIReadEvent::REALTIME_MESSAGE)
    {
        usbMidi.send(parser.getRealTimeMessage());
    } 
    
    // MIDI Controllers should discard incoming MIDI messages.
    // http://forum.pjrc.com/threads/24179-Teensy-3-Ableton-Analog-CC-causes-midi-crash
    // update reads the messages internally and discards them if we don't have any processing callbacks set
    usbMidi.update(); 

```

It looks naive but seems to work well enough. If there are better ways to do it, I will appreciate your suggestions.

**Comment by tttapa** [comment, marked as answer]
> I'm working on a device that receives MIDI bytes from another device as raw bytes and has to pass them to USB MIDI as actual MIDI messages. The source is not serial, BLE or any other normally MIDI-compatible input port, therefore Control Surface does not have interfaces for that.

Raw MIDI bytes are not really a thing, as far as I'm aware. The representation of MIDI messages depends on the transport being used, e.g. a serial connection as in the original spec, or a USB connection as in the MIDI over USB 1.0 and 2.0 specs. You cannot just transmit the bytes from one type of transport to a different type of transport, you'll usually have to parse the messages and encapsulate them into suitable packets.

Assuming that by “raw MIDI bytes” you mean “MIDI data in the serial/UART data format described by the original MIDI 1.0 spec”, then you can indeed use the `SerialMIDI_Parser` class to parse it into messages that can be sent over USB.

> As I see, USBMIDI_Interface does not have any kind of send function that would be able to collect raw bytes until they form a complete MIDI message and only then send it to USB.

Indeed, the smallest unit of data you can send over USB MIDI is the “32-bit USB-MIDI Event Packet”, you cannot send data byte-by-byte.

You can send raw event packets using `midi.backend.write(packet);`, where `packet` is of type `USBMIDI_Interface::MIDIUSBPacket_t`, usually `Array<uint8_t, 4>`.

> Also, I'm confused about begin(), loop() and update() calls in the examples. Some of them call Control_Surface.begin() and Control_Surface.loop(). But some call usbMidi.begin() and usbMidi.update() instead.

`Control_Surface.{begin,loop}` are high-level utility functions that internally call the `begin()` and `update()` methods of the MIDI interfaces (and many other objects) in your code.  
`Control_Surface.{begin,loop}` are convenient if you're using the high-level MIDI controller APIs, displays, MIDI elements, etc., but are overkill if you just need to update one or two MIDI interfaces.

To initialize and update MIDI interfaces, you have three main options, from high-level to low-level:
1. `Control_Surface.begin()` / `Control_Surface.loop()`: initialize and update all parts of the high-level Control Surface API, including all MIDI interfaces
2. `MIDI_Interface::beginAll()` / `MIDI_Interface::updateAll()`: initialize and update all MIDI interfaces
3. `midi.begin()` / `midi.update()`: initialize and update a specific MIDI interface

> If there are better ways to do it, I will appreciate your suggestions.

Using the latest `main` version (which I would recommend over 1.2.0), you could use something like this:
```cpp
#include <Control_Surface.h>
#include <MIDI_Parsers/BufferPuller.hpp>

USBMIDI_Interface midi;
SerialMIDI_Parser parser;

void setup() {
  midi.begin();
}

void handleEvent(MIDIReadEvent event) {
  switch (event) {
    case MIDIReadEvent::CHANNEL_MESSAGE:   midi.send(parser.getChannelMessage()); break;
    case MIDIReadEvent::SYSEX_CHUNK:       // fallthrough
    case MIDIReadEvent::SYSEX_MESSAGE:     midi.send(parser.getSysExMessage()); break;
    case MIDIReadEvent::SYSCOMMON_MESSAGE: midi.send(parser.getSysCommonMessage()); break;
    case MIDIReadEvent::REALTIME_MESSAGE:  midi.send(parser.getRealTimeMessage()); break;
    case MIDIReadEvent::NO_MESSAGE: break;
    default: break;
  }
}

uint8_t testdata[] = {0x9A, 0x10, 0x11, 0xF8, 0x12, 0x13, 0xF2, 0x55, 0x66, 0xF0, 0x14, 0xF7};

void example() {
  const uint8_t *data = testdata;
  size_t datalen = sizeof(testdata);
  auto puller = BufferPuller(data, datalen);
  MIDIReadEvent event = parser.pull(puller);
  while (event != MIDIReadEvent::NO_MESSAGE) {
    handleEvent(event);    
    event = parser.pull(puller);
  }
}

void loop() {
  static Timer<millis> timer {5000};
  if (timer)
    example();
  midi.update();
}
```

**Reply by progmars** [reply]
Thank you a lot for the clean example. Yes, with "raw" I meant the old serial MIDI format where status and channel are combined into one byte and channel messages come in 3 bytes (or less, if using running status). I was not aware that this format is so protocol-specific; I imagined that it's "the core" and all the modern protocols just expand it adding more bytes but not changing anything about this "status data data" byte rule.

**Reply by Kmanggaard** [reply]
WOW!

Wish I'd known about this earlier. I spent a lot of time hacking out a SysEx parser to extract text, but of course it already existed. I'll open a new thread.
